Description: tcg: fix memory corruption during live-migration
Author: Pavel Boldin <pboldin@mirantis.com>
Origin: https://launchpadlibrarian.net/216763913/tcg-commit-on-log-global-start.patch
Bug: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1493049
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1493049
Last-Update: 20150917

Index: qemu/exec.c
===================================================================
--- qemu.orig/exec.c
+++ qemu/exec.c
@@ -522,6 +522,7 @@ void tcg_cpu_address_space_init(CPUState
         cpu->tcg_as_listener = g_new0(MemoryListener, 1);
     }
     cpu->tcg_as_listener->commit = tcg_commit;
+    cpu->tcg_as_listener->log_global_start = tcg_commit;
     memory_listener_register(cpu->tcg_as_listener, as);
 }
 #endif
